### DISCUSSION TITLE
How to make it work with Pipewire?

### DISCUSSION THREAD

**Discussion by gerroon** [snapcast/snapcast#1108] [Q&A] [answered]
Hi

I have been using the SC server with Pulse for some years. I moved to Pipewire 9with Pulse support) and now the current fifo-sink method that I have been using does not seem to work.

What is the proper method to get the server to be used with Pipewire?

thanks

**Comment by gerroon** [comment]
No one? :(

**Comment by snootched** [comment]
If your pipewire integrates with PA,. then you can put this in your /etc/pulse/default.pa file to pipe the audio out to fifo

#sink to snapfifo for snapserver
load-module module-pipe-sink file=/tmp/snapfifo sink_name=Snapcast format=s16le rate=48000

I just set this up today on a VM that's getting bluetooth audio and sending it into the snapcast server (BT -> PW / PA dummy -> fifo -> snapcast)  Seems to work.  I don't have much experience with PW (yet) though.

**Comment by grmrgecko** [comment, marked as answer]
```
# cat /etc/pipewire/pipewire.conf.d/snapcast.conf
context.modules = [
{   name = libpipewire-module-pipe-tunnel
    args = {
        node.name="snapcast"
        tunnel.mode = sink
        #tunnel.may-pause = true
        # Set the pipe name to tunnel to
        pipe.filename = "/tmp/snapfifo"
        audio.format=S16LE
        audio.rate=48000
        #audio.channels=<number of channels>
        #audio.position=<channel map>
        #target.object=<remote target node>
        stream.props = {
            # extra sink properties
        }
    }
}
]
```

**Reply by igbjnI05bF** [reply]
The above looks like it's for a snapcast server run locally on your system, but what about a snapcast server run on the network? I know there is a snapcast-discover module in pipewire now and I'm having a hard time finding good documentation for it.

**Reply by grmrgecko** [reply]
Reading https://docs.pipewire.org/page_module_snapcast_discover.html I'd think you'd do similar to what I did above. Looks like you can either set the IP of the snapcast server, or have it discover when you set the IP to `~.*`.

**Reply by igbjnI05bF** [reply]
Ok and final question, not sure if you'd know the answer - but when you do this it seems like there is no stream source added to snapcast by default, only a temporary one when you are actually playing audio, so it seems like I can't add this to the "meta" stream in snapcast. If I try to add it to meta stream (meta:///mpd/mopidy/desktop-pipewire for example) then restart snapcast, it will fail as desktop-pipewire isn't a valid source stream yet.

The minor issue with this is then that you have to remember to go to your snapweb interface and manually change the source to desktop-pipewire, and then change it back to meta when you're done.

Is there a better solution for this currently? I can think of two solutions for this - have snapcast not fail to start if it isn't currently aware of a named source stream, knowing that named source streams from pipewire may pop up temporarily. or, somehow allow for a wildcard in meta, like pipewire*, which would match all pipewire streams.

**Reply by gerroon** [reply]
How does a local snapclient play from this? Running `snapclient` doesn't play the stream :(

Do I also need to modify snapserver.conf and snapclient.conf?

> ```
> # cat /etc/pipewire/pipewire.conf.d/snapcast.conf
> context.modules = [
> {   name = libpipewire-module-pipe-tunnel
>     args = {
>         node.name="snapcast"
>         tunnel.mode = sink
>         #tunnel.may-pause = true
>         # Set the pipe name to tunnel to
>         pipe.filename = "/tmp/snapfifo"
>         audio.format=S16LE
>         audio.rate=48000
>         #audio.channels=<number of channels>
>         #audio.position=<channel map>
>         #target.object=<remote target node>
>         stream.props = {
>             # extra sink properties
>         }
>     }
> }
> ]
> ```

**Reply by grmrgecko** [reply]
I'm not an expert at snapcast, I just have been following documentation at https://docs.pipewire.org/page_module_pipe_tunnel.html and https://docs.pipewire.org/page_module_snapcast_discover.html - what I did works for me, I have the following source in snapcast server config:

```
source = pipe:///tmp/snapfifo?name=default
```

**Reply by gerroon** [reply]
Thanks that works
